chain/chain/src/block_processing_utils.rs

```diff
@@ -18,6 +18,17 @@ pub(crate) const MAX_PROCESSING_BLOCKS: usize = 5;
 
 /// Contains information from preprocessing a block
 pub(crate) struct BlockPreprocessInfo {
+    /// This field has two related but actually different meanings. For the first block of an
+    /// epoch, this will be set to false if we need to download state for shards we'll track in
+    /// the future but don't track currently. This implies the first meaning, which is that if
+    /// this is true, then we are ready to apply all chunks and update flat state for shards
+    /// we'll track in this and the next epoch. This comes into play when we decide what ApplyChunksMode
+    /// to pass to Chain::apply_chunks_preprocessing().
+    /// The other meaning is that the catchup code should process this block. When the state sync sync_hash
+    /// is the first block of the epoch, these two meanings are the same. But if the sync_hash is moved forward
+    /// in order to sync the current epoch's state instead of last epoch's, this field being false no longer implies
+    /// that we want to apply this block during catchup, so some care is needed to ensure we start catchup at the right
+    /// point in Client::run_catchup()
     pub(crate) is_caught_up: bool,
```

Comment:
is it possible to split this field into multiple fields (or enum) to differentiate these meanings? it feels like the field being false indicates both we want to apply the chunks and not apply the chunks based on other state such as sync_hash.

Reply:
I think actually in both cases if this field is false, we don't want to apply the chunks for shards we don't currently track, and this logic should be the same: https://github.com/near/nearcore/blob/82bd46b0dba87e693a54bb4b066ccd78ab183dc1/chain/chain/src/chain.rs#L3980

I think we probably could split it, but it's a little bit tricky. Let me think about it actually... For now in this PR it is kept as is to not have to touch too many things and possibly break something. the tricky part is that right now we add the first block of the epoch to the `BlocksToCatchup` column based on this field, which is then read to see if we'll need to catch up the next block after this one as well: https://github.com/near/nearcore/blob/82bd46b0dba87e693a54bb4b066ccd78ab183dc1/chain/chain/src/chain.rs#L2288

I guess where that is called maybe we can just call `get_state_sync_info()` again, and also check if catchup is already done, but it requires some care